sdk/servicebus/service-bus/test/internal/batchingReceiver.spec.ts

```diff
@@ -116,4 +127,153 @@ describe("BatchingReceiver unit tests", () => {
+      let clock: ReturnType<typeof sinon.useFakeTimers>;
+
+      beforeEach(() => {
+        clock = sinon.useFakeTimers();
+      });
+
+      afterEach(() => {
+        clock.restore();
+      });
+
+      it("1. We received 'max messages'", async () => {
+        const receiver = new BatchingReceiver(createClientEntityContextForTests(), {
+          receiveMode: lockMode
+        });
+
+        receiver["_getServiceBusMessage"] = (eventContext) => {
+          return {
+            body: eventContext.message?.body
+          } as ServiceBusMessageImpl;
+        };
+
+        const { receiveIsReady, emitter } = setupFakeReceiver(receiver);
+
+        const receivePromise = receiver.receive(1, bigTimeout, bigTimeout);
+        await receiveIsReady;
+
+        // batch fulfillment is checked when we receive a message...
+        emitter.emit(ReceiverEvents.message, {
+          message: { body: "the message" } as RheaMessage
+        } as EventContext);
+
+        const messages = await receivePromise;
+        assert.deepEqual(
+          messages.map((m) => m.body),
+          ["the message"]
+        );
+      }).timeout(5 * 1000);
+
+      // in the new world the overall timeout firing means we've received _no_ messages
+      // because otherwise it'd be one of the others.
+      it("2. We've waited 'max wait time'", async () => {
+        const receiver = new BatchingReceiver(createClientEntityContextForTests(), {
+          receiveMode: lockMode
+        });
+
+        const { receiveIsReady } = setupFakeReceiver(receiver);
+
+        const receivePromise = receiver.receive(1, littleTimeout, bigTimeout);
+
+        // force the overall timeout to fire
+        clock.tick(littleTimeout);
+
+        await receiveIsReady;
+        const messages = await receivePromise;
+        assert.isEmpty(messages);
+      }).timeout(5 * 1000);
+
+      it(`3. We've received 1 message and _now_ have exceeded 'max wait time past first message`, async () => {
+        const receiver = new BatchingReceiver(createClientEntityContextForTests(), {
+          receiveMode: lockMode
+        });
+
+        receiver["_getServiceBusMessage"] = (eventContext) => {
+          return {
+            body: eventContext.message?.body
+          } as ServiceBusMessageImpl;
+        };
+
+        const { receiveIsReady, emitter } = setupFakeReceiver(receiver);
+
+        const receivePromise = receiver.receive(3, bigTimeout, littleTimeout);
+        await receiveIsReady;
+
+        // batch fulfillment is checked when we receive a message...
+        emitter.emit(ReceiverEvents.message, {
+          message: { body: "the first message" } as RheaMessage
+        } as EventContext);
+
+        // advance the timeout to _just_ before the expiration of the first one (which must have been set
+        // since we just received a message). This'll make it more obvious if I scheduled it a second time.
+        clock.tick(littleTimeout - 1);
```

Comment:
I do like the use of sinon clocks :smile:

Reply:
I have some other tests that could benefit from them as well but I'll do those in another PR.